src/HealthCheck.Core/HealthCheck.cs

```diff
@@ -2,55 +2,53 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
+using HealthCheck.Core.Configuration;
+using HealthCheck.Core.Results;
 
 namespace HealthCheck.Core
 {
-    public class HealthCheck
+    public class HealthCheck : HealthCheckBase
     {
-        private readonly IEnumerable<IChecker> _checkers;
+        public HealthCheck(IEnumerable<IChecker> checkers) : base(checkers) { }
 
-        public HealthCheck(IEnumerable<IChecker> checkers)
-        {
-            _checkers = checkers;
-        }
-
-        public async Task<HealthCheckResult> Run()
+        public override async Task<HealthCheckResult> Run()
         {
             try
             {
-                if (_checkers == null)
+                if (Checkers == null)
                 {
-                    return new HealthCheckResult(Enumerable.Empty<CheckResult>());
+                    return new HealthCheckResult(Enumerable.Empty<ICheckResult>());
                 }
                 // Run all checks in parallel (if possible). Catch exceptions for each
                 // checker individually.
-                var checkResults = _checkers
-                    .AsParallel()
+                var checkResults = Checkers.AsParallel()
                     .Select(async x =>
                     {
                         try
                         {
-                            return await x.Check().ConfigureAwait(false);
+                            return !x.PreserveContext ? await x.Check().ConfigureAwait(false) : x.Check().Result;
```

Comment:
Because we have faced situation when we do need capturing context. If i am not mistaking when using it from mvc or web api controller we are loosing web context which used in one particular healthcheck.

Reply:
I don't think it's a good idea to use for example `HttpContext.Current` in a checker implementation either way, from what I've read it's not designed for being used from different threads... That being said, I'm leaning towards skipping `ConfigureAwait` alltogether instead.

Reply:
Well... you have created universal tool for implementing healthchecks. Healthcheck term is so abstract it can be anything. Even value from current web context for particular project. You cannot restrict people in capabilities of your tool - they just wont use it right?
Do you want concrete example? - We have been implementing some healhcheck for umbraco based project. For reusing some of built-in healthchecks available only in ui for our web api call we NEEDED current context.

Reply:
Ok then.

Could you change that line into this though:
```
return !x.PreserveContext ? await x.Check().ConfigureAwait(false) : await x.Check();
```

Reply:
I suppose i have confused you a bit. As far as i remember (it was 6 month ago) we exactly needed to run checker at the same thread to have an access to WebContext. That s why i "awaited" task but asked for .Result. I made a change you required i suppose guys will report here if it will be an issue again. 

@minorum  Yuo should check it guys in your projects. If your healthcheck will become broken we will need to redesign this change back.

Reply:
I see...it still seems a bit strange to me though; the `HttpContext.Current` reflects the current request, which in all cases for the healthcheck endpoint would probably be initiated by some kind of monitoring or automatic test system. What request-dependant properties are you performing health checks on? (I'm not saying that it's wrong, I'm just curious...)